### DISCUSSION TITLE
Difference between dctur, relay server, rendezvous protocol, signalling server and a tracking server?

### DISCUSSION THREAD

**Discussion by souvikinator** [libp2p/rust-libp2p#5357] [Q&A] [answered]
Hi, I was able to make a p2p network discover new nodes in the local network and were able to publish messages using gossipsub protocol and mDNS.

Now I want to discover nodes in a public network and I was going through the examples in the repo and went through a few:
- dcutr
- rendezvous
- autoNAT
- relay server

and while it feels like they are solving similar problem so I wanted to understand the difference between the
relay server, rendezvous protocol, signaling server, and a tracking server? and when to use what?

**Comment by dariusc93** [comment, marked as answer]
In short:
1. DCUtR (Direct Connection Upgrade through Relay) is a form of hole punching that allows establishing a connection to a peer directly through a relay. If a direct connection fails (which there are many reasons on why it would), the local node will fall back to the relay). This does require `libp2p-relay` (or specifically HOP protocol) to use this protocol.
2. Rendezvous (protocol) is used for peer discovery. This is really more application specific but does allow one to use for discovering peers (that have registered) in a given namespace. Can be used for bootstrapping, discovering peers to a specific public topic, content discovery, etc., though still more application specific. 
3. AutoNAT detects if you are behind a NAT by sending a request out to a public node to have your local node be dialed by the address(es) that your node provided. 
4. Relay server (v2) allows you to establish a connection to a peer through a relay as well as allowing peers to connect to you through said relay. 


If you want to discover nodes in a public network, the network would be using some protocol for that. Commonly, kad is used for this, but discovery could also happen through pubsub, or some other protocol.
